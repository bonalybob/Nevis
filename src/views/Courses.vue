<template>
  <main>
    <back-arrow :to="`/events/${$route.params.id}`" />
    <h1 class="mx-10 mb-1 px-1">
      Courses
    </h1>
    <div class="mx-12 mb-3">
      <router-link
        :to="`/events/${$route.params.id}/courses/create`"
        tag="button"
        class="button"
      >
        Create Course
      </router-link>
      <button @click="getCourses" class="button">Refresh</button>
    </div>
    <template v-if="courses && courses.length > 0">
      <router-link
        v-for="course of courses"
        :key="course.id"
        :to="`/events/${$route.params.id}/courses/${course.id}/edit`"
        tag="div"
        class="my-shadow mx-12 mb-5 px-3 py-2 border-t-4 border-blue"
      >
        <h2 class="pb-1">{{ course.name }}</h2>
        <p>
          <b>Length:</b>
          {{ course.length / 1000 }} km
          <b />
          <b>Climb:</b>
          {{ course.climb }} m
        </p>
        <p>
          <b>Controls:</b>
          {{ course.controls }}
        </p>
        <p>
          <b>Number of Entries:</b>
          {{ course.numberOfCompetitors }}
        </p>
      </router-link>
    </template>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'

export default {
  components: {
    'back-arrow': BackArrow,
  },

  data: function() {
    return {
      courses: [],
    }
  },

  mounted: function() {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    } else this.getCourses()
  },

  methods: {
    getCourses: function() {
      return this.$database
        .query(
          `
      SELECT courses.*, IFNULL(COUNT(competitors.course),0) AS numberOfCompetitors
      FROM courses
      LEFT JOIN competitors ON courses.id=competitors.course
      WHERE courses.event=?
      GROUP BY courses.id
      ORDER BY courses.name`,
          this.$route.params.id
        )
        .then(result => {
          this.courses = result
        })
        .catch(() =>
          this.$messages.addMessage('Problem Fetching Courses', 'error')
        )
    },
  },
}
</script>
