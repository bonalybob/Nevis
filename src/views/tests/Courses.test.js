import { mount, shallowMount } from '@vue/test-utils'
import Courses from '@/views/Courses'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - No Courses', () => {
  const wrapper = shallowMount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - With Courses', () => {
  const wrapper = shallowMount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({
    courses: [
      { id: 0, name: 'Long', length: 2.35, climb: 150, controls: '101,102,103' },
      { id: 77, name: 'Short', length: 1.75, climb: 15, controls: '100,108,103' },
    ],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: false, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Courses - Success', async () => {
  const wrapper = mount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue(['hello']) },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.courses).toEqual(['hello'])
})

test('Get Courses - Error', async () => {
  const wrapper = mount(Courses, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Courses', 'error')
  expect(wrapper.vm.courses.length).toBe(0)
})
