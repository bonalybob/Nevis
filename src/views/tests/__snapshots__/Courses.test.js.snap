// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - No Courses 1`] = `
<main>
  <back-arrow-stub
    to="/events/12"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    
    Courses
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/12/courses/create"
    >
      
      Create Course
    
    </router-link-stub>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <!---->
</main>
`;

exports[`Renders Correctly - With Courses 1`] = `
<main>
  <back-arrow-stub
    to="/events/12"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    
    Courses
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/12/courses/create"
    >
      
      Create Course
    
    </router-link-stub>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <router-link-stub
    class="my-shadow mx-12 mb-5 px-3 py-2 border-t-4 border-blue"
    tag="div"
    to="/events/12/courses/0/edit"
  >
    <h2
      class="pb-1"
    >
      Long
    </h2>
     
    <p>
      <b>
        Length:
      </b>
      
        0.00235 km
        
      <b />
       
      <b>
        Climb:
      </b>
      
        150 m
      
    </p>
     
    <p>
      <b>
        Controls:
      </b>
      
        101,102,103
      
    </p>
     
    <p>
      <b>
        Number of Entries:
      </b>
      
        
      
    </p>
  </router-link-stub>
  <router-link-stub
    class="my-shadow mx-12 mb-5 px-3 py-2 border-t-4 border-blue"
    tag="div"
    to="/events/12/courses/77/edit"
  >
    <h2
      class="pb-1"
    >
      Short
    </h2>
     
    <p>
      <b>
        Length:
      </b>
      
        0.00175 km
        
      <b />
       
      <b>
        Climb:
      </b>
      
        15 m
      
    </p>
     
    <p>
      <b>
        Controls:
      </b>
      
        100,108,103
      
    </p>
     
    <p>
      <b>
        Number of Entries:
      </b>
      
        
      
    </p>
  </router-link-stub>
</main>
`;
